---
title: "R Notebook"
output: 
html_document:
    toc: True
    toc_float: True
    df_print: paged
---

```{r}
#load libraries
library(tidyverse)
```

```{r}
#loaded R file from leafcutter
load("/Volumes/My_Passport/Transcriptomics/Biogen_LRKK2/Biogen_LRRK2.RData")

#load sample data
Biogen_LRRK2_leafcutter_ds_cluster_significance <- read.delim("/Volumes/My_Passport/Transcriptomics/Biogen_LRKK2/leafcutter_ds_cluster_significance.txt")

```

```{r}
#separate columns to highlight cluster ID
Biogen_LRRK2_leafcutter_ds_cluster_significance_sorted <- Biogen_LRRK2_leafcutter_ds_cluster_significance %>%
  separate('cluster', c('Chromosome','Cluster_ID'),':') 

clusters <- rename(clusters, Cluster_ID=clusterID) %>%
  select(-gene)

gene_names <- select(introns,clusterID,gene) %>%
  rename(Cluster_ID=clusterID) %>%
  distinct() %>%
  filter(gene!= ".")

```

```{r}
#merge only tables needed together to get list of gene names + FDR + Pvalue

Biogen_LRRK2_final <- select(clusters, Cluster_ID, FDR)
Biogen_LRRK2_PValue <- select(Biogen_LRRK2_leafcutter_ds_cluster_significance_sorted, Cluster_ID, p.adjust)

Biogen_LRRK2_list <- inner_join(gene_names, Biogen_LRRK2_PValue, by=c('Cluster_ID'))  %>%
  inner_join(Biogen_LRRK2_final, by=c('Cluster_ID')) %>%
  select(gene,p.adjust,FDR) %>%
  filter(FDR<= 0.05 ) %>%
  drop_na() %>%
  group_by( gene) %>%
  arrange(FDR, .by_group=T) %>% #groups by Gene name and sorts them by FDR
  top_n(-1, FDR) %>% # selects for each gene the pone with lowest FDR
  distinct()


#Save
write.csv(Biogen_LRRK2_list,'Biogen_LRRK2_LeafCutter_Clean_0.05.csv', row.names = F)

```



